spring-core/src/main/java/org/springframework/util/MimeTypeUtils.java

```diff
@@ -420,54 +421,71 @@ public static String generateMultipartBoundaryString() {
 
 		private final ConcurrentHashMap<K, V> cache = new ConcurrentHashMap<>();
 
-		private final ReadWriteLock lock = new ReentrantReadWriteLock();
+		private final Lock readLock;
+
+		private final Lock writeLock;
 
 		private final Function<K, V> generator;
 
+		private volatile int size = 0;
+
 		public ConcurrentLruCache(int maxSize, Function<K, V> generator) {
 			Assert.isTrue(maxSize > 0, "LRU max size should be positive");
 			Assert.notNull(generator, "Generator function should not be null");
 			this.maxSize = maxSize;
 			this.generator = generator;
+
+			ReadWriteLock lock = new ReentrantReadWriteLock();
+			this.readLock = lock.readLock();
+			this.writeLock = lock.writeLock();
 		}
 
 		public V get(K key) {
-			this.lock.readLock().lock();
-			try {
-				if (this.queue.size() < this.maxSize / 2) {
-					V cached = this.cache.get(key);
-					if (cached != null) {
-						return cached;
-					}
+			V cached;
+
+			if ((cached = this.cache.get(key)) != null) {
+				if (this.size < this.maxSize / 2) {
+					return cached;
 				}
-				else if (this.queue.remove(key)) {
+
+				try {
+					this.readLock.lock();
 					this.queue.add(key);
-					return this.cache.get(key);
+					this.queue.remove(key);
+					return cached;
+				}
+				finally {
+					this.readLock.unlock();
 				}
 			}
-			finally {
-				this.lock.readLock().unlock();
-			}
-			this.lock.writeLock().lock();
+
+			this.writeLock.lock();
 			try {
 				// retrying in case of concurrent reads on the same key
-				if (this.queue.remove(key)) {
+				if ((cached = this.cache.get(key)) != null) {
 					this.queue.add(key);
-					return this.cache.get(key);
+					this.queue.remove(key);
+					return cached;
 				}
-				if (this.queue.size() == this.maxSize) {
+
+				int cacheSize = this.size;
+				if (cacheSize == this.maxSize) {
```

Comment:
Recommended to use >= instead of ==

Reply:
I don't understand how is it possible to occur a problem with write lock.

Reply:
Reordered `V value = this.generator.apply(key);` line, To prevent size inconsistency in case generator throws any exception.

Reply:
OK, Understand